### ISSUE TITLE
ember-data date transform: triggering change on record re-insertion

### ISSUE THREAD

**Issue by pankaj-awake** [emberjs/data#9242]
### Reproduction

I have set up a github repository demonstrating the issue.

Here is the [TestEmberDataDateTransform Repo Link](https://github.com/pankaj-awake/ember-test-date-transform)

Once you clone and start the application:

- A date attribute 'createdAt' is logged every time the 'Reload record' button is clicked. 
- The non-date attributes "name" and "isFake" are just logged on the initial load.
- View the console for the output of the change event.

### Description

The issue at play seems to happen when using the date transform from version 4.12.0 onwards. The date transform triggers a change on the attribute whenever the record is re-inserted into the store, even when the value of the concerned attribute remains the same. This behaviour appears inconsistent when compared with non-date attributes of the same record. The _expected_ result would be that all attributes behave consistently on reload, but the _unexpected_ result is the change that is triggered on 'date' data type.

In regards to the Ember-data version, the most recent version that worked as `expected` is version 4.11.3.

### Versions

I have used the following versions in my ember application:
```
yarn list v1.22.19
└─ ember-source@4.12.4
✨  Done in 0.23s.
yarn list v1.22.19
├─ @glimmer/component@1.1.2
│  └─ ember-cli-version-checker@3.1.3
├─ ember-cli-app-version@6.0.1
├─ ember-cli-babel-plugin-helpers@1.1.1
├─ ember-cli-babel@7.26.11
│  └─ ember-cli-version-checker@4.1.1
├─ ember-cli-dependency-checker@3.3.2
├─ ember-cli-get-component-path-option@1.0.0
├─ ember-cli-htmlbars@6.3.0
├─ ember-cli-inject-live-reload@2.1.0
│  └─ ember-cli-version-checker@3.1.3
├─ ember-cli-is-package-missing@1.0.0
├─ ember-cli-lodash-subset@2.0.1
├─ ember-cli-normalize-entity-name@1.0.0
├─ ember-cli-path-utils@1.0.0
├─ ember-cli-preprocess-registry@3.3.0
├─ ember-cli-sri@2.1.1
├─ ember-cli-string-utils@1.1.0
├─ ember-cli-terser@4.0.2
├─ ember-cli-test-info@1.0.0
├─ ember-cli-test-loader@3.1.0
├─ ember-cli-typescript-blueprint-polyfill@0.1.0
├─ ember-cli-typescript@3.0.0
├─ ember-cli-version-checker@5.1.2
├─ ember-cli@4.12.2
├─ ember-fetch@8.1.2
│  └─ ember-cli-typescript@4.2.1
├─ ember-load-initializers@2.1.2
│  └─ ember-cli-typescript@2.0.2
└─ ember-tracked-storage-polyfill@1.0.0
   └─ ember-cli-htmlbars@5.7.2
✨  Done in 0.24s.
yarn list v1.22.19
├─ @ember-data/adapter@4.12.5
├─ @ember-data/debug@4.12.5
├─ @ember-data/graph@4.12.5
├─ @ember-data/json-api@4.12.5
├─ @ember-data/legacy-compat@4.12.5
├─ @ember-data/model@4.12.5
├─ @ember-data/private-build-infra@4.12.5
├─ @ember-data/request@4.12.5
├─ @ember-data/rfc395-data@0.0.4
├─ @ember-data/serializer@4.12.5
├─ @ember-data/store@4.12.5
├─ @ember-data/tracking@4.12.5
├─ babel-plugin-ember-data-packages-polyfill@0.1.2
└─ ember-data@4.12.5
```

Thanks for your contributions in making Ember.js more robust!

Regards, 
Pankaj Kataria

**Comment by runspired** [comment]
this is expected and has always been the case. There is no potential fix for this in the `@ember-data/model` paradigm. The model replacement that is under development won't provide any transforms out of the box, but its pattern for them would allow you to avoid this instability were you to implement one.

**Comment by runspired** [comment]
related: https://github.com/emberjs/data/issues/1540

**Comment by pankaj-awake** [comment]
Thank you for your swift and enlightening response.

I appreciate the suggestion to implement a fix for our projects, which we do intend to undertake. However, I wanted to bring to your attention that the date transform began triggering changes since the update to ember-data v12.0.0. Please note that this behaviour was not observed in the versions preceding this one. (my initial communication may not have reflected this point clearly)

Once again, thank you for your time and the insight you've provided. I (my team) looks forward to the forthcoming improvements in Version 5.4.

**Comment by runspired** [comment]
> Please note that this behaviour was not observed in the versions preceding this one. (my initial communication may not have reflected this point clearly)

This has quite literally always been the case. hence the link to the longstanding issue for 2013 🙈

You will note the problem in that issue is exactly the same, and was never resolved for `@ember-data/model`.

This is for the simple reason of *how* transforms work with @ember-data/model, or (dare I say) don't work. The typical assumption is that transforms work by converting data into/out of another format on access by the model. However, the model does not utilize transforms at all. Instead transforms are used by some (not all) serializers to transform data between the api's format and the cache's format.

What this means is that the `date` transform, when used, has always been a leaky design. It results in the cache containing a `Date` instance. Date instances are both mutable and two instances with an identical date are not strictly equal. This means any time data is received by the cache as a Date instance it will fail a simple `===` check and result in a notification being sent to the model indicating the field has changed.

Why this did not occur for you in 4.11 is anyone's guess, my suspicion would be that somewhere in your app is some code that in older versions would have worked around this limitation for you, perhaps with private api usage that no longer works.

**Comment by runspired** [comment]
After digging some more, a year after this issue was reported here, Ember changed `isEqual` to compare date instances using `getTime`, then in 2015 a change got slid in to make the `===` comparison use `isEqual`. It was never explicitly to support dates and definitely didn't fix the underlying issue that any two objects that don't have strict identity will trigger notifications. In 4.12 as we worked to remove more of the unnecessary entanglement with Ember, we stopped using `isEqual`. Since date comparison was not one of the widely known features of `isEqual` this again went unnoticed by any except those like yourself that became dependent on it.

Generally speaking, such a notification in these cases is both safer and more desirable. In a pull base reactivity world, its unsafe to assume a non-strict comparison should go silent.
